samza-core/src/main/scala/org/apache/samza/container/TaskInstance.scala

```diff
@@ -128,46 +151,52 @@ class TaskInstance(
     debug("Registering consumers for taskName: %s" format taskName)
 
     systemStreamPartitions.foreach(systemStreamPartition => {
-      val offset = offsetManager.getStartingOffset(taskName, systemStreamPartition)
-      .getOrElse(throw new SamzaException("No offset defined for SystemStreamPartition: %s" format systemStreamPartition))
-      consumerMultiplexer.register(systemStreamPartition, offset)
-      metrics.addOffsetGauge(systemStreamPartition, () => {
-        offsetManager
-          .getLastProcessedOffset(taskName, systemStreamPartition)
-          .orNull
-      })
+      val startingOffset = getStartingOffset(systemStreamPartition)
+      consumerMultiplexer.register(systemStreamPartition, startingOffset)
+      metrics.addOffsetGauge(systemStreamPartition, () =>
+        if (sideInputSSPs.contains(systemStreamPartition)) {
+          sideInputStorageManager.getLastProcessedOffset(systemStreamPartition)
+        } else {
+          offsetManager.getLastProcessedOffset(taskName, systemStreamPartition).orNull
+        })
     })
   }
 
   def process(envelope: IncomingMessageEnvelope, coordinator: ReadableCoordinator,
     callbackFactory: TaskCallbackFactory = null) {
     metrics.processes.inc
 
-    if (!ssp2CaughtupMapping.getOrElse(envelope.getSystemStreamPartition,
-      throw new SamzaException(envelope.getSystemStreamPartition + " is not registered!"))) {
+    val incomingMessageSsp = envelope.getSystemStreamPartition
+
+    if (!ssp2CaughtupMapping.getOrElse(incomingMessageSsp,
+      throw new SamzaException(incomingMessageSsp + " is not registered!"))) {
       checkCaughtUp(envelope)
     }
 
-    if (ssp2CaughtupMapping(envelope.getSystemStreamPartition)) {
+    if (ssp2CaughtupMapping(incomingMessageSsp)) {
       metrics.messagesActuallyProcessed.inc
 
       trace("Processing incoming message envelope for taskName and SSP: %s, %s"
-        format (taskName, envelope.getSystemStreamPartition))
+        format (taskName, incomingMessageSsp))
 
-      if (isAsyncTask) {
-        exceptionHandler.maybeHandle {
-          val callback = callbackFactory.createCallback()
-          task.asInstanceOf[AsyncStreamTask].processAsync(envelope, collector, coordinator, callback)
-        }
+      if (sideInputSSPs.contains(incomingMessageSsp)) {
+        sideInputStorageManager.process(envelope)
       } else {
-        exceptionHandler.maybeHandle {
-         task.asInstanceOf[StreamTask].process(envelope, collector, coordinator)
-        }
+        if (isAsyncTask) {
```

Comment:
Just curious.

* AsyncStreamTask numProcess calls: By existing AsynStreamTask semantics, `task.max.concurrency` denotes the maximum number of `TaskInstance.process` calls that can happen  at a time. Now after this change, we're combining SideInput.process and TaskInstance.process together and that will affect the overall processing throughput of  input streams. Since anyways after bootstrap, we want the store update happen in background in changelog, do we want SideInput.process to affect the number of task process calls? In worst case, sideInput message arrival rate could throttle the message processing rate of input streams of the job.  
* StreamTask commit and process: By existing semantics TaskInstance.commit and TaskInstance.process cannot happen at same time for StreamTask. Now after this change after bootstrap, SideInputProcess.process and TaskInstance.commit cannot occur at same time. Similarily sideInput.process and taskInstance.process cannot happen at same time.   Ideally after sideinput bootstrap, SideInput.process() should not affect the normal job's throughput (occurence of task.process() and task.commit()). 

Please share your thoughts.

Even if we really want to go down this route then atleast it would be better to document it .

Reply:
These are good points, thanks for bringing them up.

For 1, we'll see how this new feature performs and tune accordingly. If task throughput is affected significantly, we have the option of taking side input consumption out of the run loop and into its own thread(s) with its own consumer(s), similar to changelog restores.

For 2, SideInput#process is currently a part of TaskInstance.process, they're not exclusive. It is only exclusive with Task.process for non-async processing. We can document these semantics in more detail in public documentation when we have a stable design.

Please consider this PR an alpha version for experimentation, not the final release. @bharathkk For this reason, I'd prefer to _not_ document the feature yet, including adding it to configuration reference. Also, all public APIs should have the appropriate InterfaceStability annotations.